router/router.go

```diff
@@ -1849,3 +1860,66 @@ func (rt *HandleT) Resume() {
 
 	rt.paused = false
 }
+
+// Currently the retry logic has been implemented for only OAuth
+func (rt *HandleT) SendToTransformerProxyWithRetry(val integrations.PostParametersT,
+	destinationJob types.DestinationJobT, accessToken string, retryCount int) (statusCode int, response string) {
+	var respStatusCode int
+	var respBodyTemp string
+
+	respStatusCode, respBodyTemp = rt.transformer.Send(val, rt.destName, accessToken)
+	if retryCount >= 2 {
+		// Retrial termination condition
+		return respStatusCode, respBodyTemp
+	}
+	fmt.Printf("SendToProxy Response Code: %d\n", respStatusCode)
+	fmt.Printf("SendToProxy Response: %s\n", respBodyTemp)
+	var errOutput oauth.ErrorOutput
+	if err := json.Unmarshal([]byte(respBodyTemp), &errOutput); err != nil {
+		// If respBodyTemp comes out with a string, then this will occur
+		// TODO: Should we send any other error here ?
+		return http.StatusInternalServerError, err.Error()
+	}
+	workspaceId := destinationJob.JobMetadataArray[0].WorkspaceId
+	var stCd int = 0
+	var res string = ""
+	// Check the category
+	// Trigger the refresh endpoint/disable endpoint
+	if errOutput.Output.AuthErrorCategory == oauth.DISABLE_DEST {
+		stCd, res = rt.oauth.DisableDestination(destinationJob.Destination, workspaceId)
+		if stCd == 200 {
+			// Abort the jobs as the destination is disable
+			return http.StatusBadRequest, res
+		}
+	} else if errOutput.Output.AuthErrorCategory == oauth.REFRESH_TOKEN {
+		accountId := destinationJob.JobMetadataArray[0].AccountId
+		stCd, res = rt.oauth.RefreshToken(workspaceId, accountId, errOutput.Output.AccessToken)
+		if stCd == 200 && router_utils.IsNotEmptyString(res) {
+			retryCount += 1
+			// Setting these values since we would need to update the cache using these values
+			val.WorkspaceId = workspaceId
+			val.AccountId = accountId
+			// This will be "true" only when a refresh executed successfully
+			var accountSecret oauth.AccountSecret
+
+			if router_utils.IsNotEmptyString(res) {
+				if err := json.Unmarshal([]byte(res), &accountSecret); err != nil {
+					// Some problem with AccountSecret unmarshalling
+					return http.StatusInternalServerError, err.Error()
+				} else if !router_utils.IsNotEmptyString(accountSecret.AccessToken) {
```

Comment:
need not check for expirationDate? What would be the case where we would be having an empty accessToken. Is it being retried on transformer side for such cases?

Reply:
There can be valid cases where a destination doesn't expire the access token at all in which case expiration date will not be available. Since this is a valid case, I've removed. that check. Let me know if it doesn't make sense